swarm/storage/hasherstore.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/crypto/sha3"
+	"github.com/ethereum/go-ethereum/swarm/storage/encryption"
+)
+
+type chunkEncryption struct {
+	spanEncryption encryption.Encryption
+	dataEncryption encryption.Encryption
+}
+
+type hasherStore struct {
+	store           ChunkStore
+	hashFunc        SwarmHasher
+	chunkEncryption *chunkEncryption
+	refSize         int64
+	wg              *sync.WaitGroup
+	closed          chan struct{}
+}
+
+func newChunkEncryption() *chunkEncryption {
+	return &chunkEncryption{
+		spanEncryption: encryption.New(0, math.MaxUint32, sha3.NewKeccak256),
+		dataEncryption: encryption.New(4096, 0, sha3.NewKeccak256),
+	}
+}
+
+// NewHasherStore creates a hasherStore object, which implements Putter and Getter interfaces.
+// With the HasherStore you can put and get chunk data (which is just []byte) into a ChunkStore
+// and the hasherStore will take core of encryption/decryption of data if necessary
+func NewHasherStore(chunkStore ChunkStore, hashFunc SwarmHasher, toEncrypt bool) *hasherStore {
```

Comment:
I think `HasherStore` is a good candidate for a few unit tests.

Reply:
I totally agree, I'm adding a few

Reply:
Unit test added in cdc7bed7ab5e9e3fa31b0f4b3a473d15f70c7b64

Reply:
@nonsense  Nice, I committed it in ba258d747e676201bde3ace976bef0432bbb8d6e